traffic_ops/testing/api_contract/v4/conftest.py

```diff
@@ -0,0 +1,122 @@
+from trafficops.restapi import OperationError
+
+
+# Create and configure logger
+logger = logging.getLogger()
+
+
+def pytest_addoption(parser: object) -> None:
+    """Passing in Traffic Ops arguments [Username, Password, Url and Hostname] from command line.
+    :param parser: Parser to parse command line arguments
+    """
+    parser.addoption(
+        "--to-user", action="store", help="User name for Traffic Ops Session."
+    )
+    parser.addoption(
+        "--to-password", action="store", help="Password for Traffic Ops Session."
+    )
+    parser.addoption(
+        "--to-url", action="store", help="Traffic Ops URL."
+    )
+
+
+@pytest.fixture(name="to_args")
+def to_data(pytestconfig: pytest.Config) -> dict:
+    """PyTest fixture to store Traffic ops arguments passed from command line.
+    :param pytestconfig: Session-scoped fixture that returns the session's pytest.Config object
+    :returns args: Return Traffic Ops arguments
+    """
+    args = {}
+    with open("to_data.json", encoding="utf-8", mode="r") as session_file:
+        data = json.load(session_file)
+    session_data = data["test"]
+    args["api_version"] = urlparse(
+        session_data.get("url")).path.strip("/").split("/")[1]
+    args["port"] = session_data.get("port")
+
+    to_user = pytestconfig.getoption("--to-user")
+    to_password = pytestconfig.getoption("--to-password")
+    to_url = pytestconfig.getoption("--to-url")
+
+    if not all([to_user, to_password, to_url]):
+        logger.info(
+            "Traffic Ops session data were not passed from Command line Args.")
+        args["user"] = session_data.get("user")
+        args["password"] = session_data.get("password")
+        args["url"] = session_data.get("url")
+    else:
+        args["user"] = to_user
+        args["password"] = to_password
+        args["url"] = to_url
+        logger.info("Parsed Traffic ops session data from args %s", args)
+    return args
+
+
+@pytest.fixture(name="to_session")
+def to_login(to_args: dict) -> TOSession:
+    """PyTest Fixture to create a Traffic Ops session from Traffic Ops Arguments
+    passed as command line arguments in to_args fixture in conftest.
+    :param to_args: Fixture to get Traffic ops session arguments
+    :returns to_session: Return Traffic ops session
+    """
+    # Create a Traffic Ops V4 session and login
+    to_url = urlparse(to_args["url"])
+    to_host = to_url.hostname
+    try:
+        to_session = TOSession(host_ip=to_host, host_port=to_args["port"],
+                               api_version=to_args["api_version"], ssl=True, verify_cert=False)
+        logger.info("Established Traffic Ops Session.")
+    except OperationError:
+        sys.exit(-1)
+
+    # Login To TO_API
+    to_session.login(to_args["user"], to_args["password"])
+    logger.info("Successfully logged into Traffic Ops.")
+    return to_session
+
+
+@pytest.fixture()
+def cdn_prereq(to_session: TOSession, get_cdn_data: dict) -> list:
```

Comment:
This can return `None` in addition to a list, so it should be annotated with `Optional`. Also, the list has a more specific type than `list[any]`, it's `list[dict[str, str] | CDN]` but idk how you wanna type a CDN. It's meant to be data read in from a file that is used to create a CDN, so you can probably make that type pretty specific.

Speaking of, though, these names confuse me. This fixture is named `cdn_prereq`, but what it returns is data from the API - in a *POST* request, too, not *GET* - and the input argument `get_cdn_data` isn't actually retrieved from the API in a *GET* request, that's the prerequisite data? The same seeming mismatch is occurring in `test_cdns.py`.

Reply:
Changed the fixture name from cdn_prereq to cdn_post_data and get_cdn_data to cdn_prereq_data in test_cdns.py
And added the return type as Optional[list[dict[str, str] | requests.Response]]